File: src/test/java/de/fraunhofer/aisec/crymlin/RuleEnsureSemanticsTest.java

package de.fraunhofer.aisec.crymlin;

import de.fraunhofer.aisec.analysis.markevaluation.ExpressionEvaluator;
import de.fraunhofer.aisec.analysis.structures.AnalysisContext;
import de.fraunhofer.aisec.analysis.structures.ConstantValue;
import de.fraunhofer.aisec.analysis.structures.MarkContextHolder;
import de.fraunhofer.aisec.analysis.structures.MarkIntermediateResult;
import de.fraunhofer.aisec.analysis.structures.ServerConfiguration;
import de.fraunhofer.aisec.analysis.structures.TypestateMode;
import de.fraunhofer.aisec.cpg.TranslationManager;
import de.fraunhofer.aisec.cpg.TranslationResult;
import de.fraunhofer.aisec.mark.XtextParser;
import de.fraunhofer.aisec.mark.markDsl.Expression;
import de.fraunhofer.aisec.mark.markDsl.MarkModel;
import de.fraunhofer.aisec.markmodel.MRule;
import de.fraunhofer.aisec.markmodel.Mark;
import de.fraunhofer.aisec.markmodel.MarkModelLoader;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

import java.io.File;
import java.net.URL;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;
import java.util.stream.Collectors;

import static de.fraunhofer.aisec.cpg.graph.GraphKt.getGraph;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.junit.jupiter.api.Assertions.fail;

class RuleEnsureSemanticsTest {

	private static HashMap<String, MarkModel> markModels;

	@BeforeAll
	static void startup() {
		URL resource = RuleEnsureSemanticsTest.class.getClassLoader().getResource("mark/rules/ensure/semantics/");
		assertNotNull(resource);

		File markFile = new File(resource.getFile());
		assertNotNull(markFile);

		File[] directoryContent = markFile.listFiles((current, name) -> name.endsWith(".mark"));

		if (directoryContent == null) {
			directoryContent = new File[] { markFile };
		}

		assertNotNull(directoryContent);
		assertTrue(directoryContent.length > 0);

		XtextParser parser = new XtextParser();
		for (File mf : directoryContent) {
			parser.addMarkFile(mf);
		}

		markModels = parser.parse();
		assertFalse(markModels.isEmpty());
	}

	private void test(String markFileEnding) {
		List<String> markFilePaths = markModels.keySet().stream().filter(n -> n.endsWith(markFileEnding)).collect(Collectors.toList());
		assertEquals(1, markFilePaths.size());

		Mark mark = new MarkModelLoader().load(markModels, markFilePaths.get(0));
		ServerConfiguration config = ServerConfiguration.builder().markFiles(markFilePaths.get(0)).typestateAnalysis(TypestateMode.NFA).build();

		var graph = getGraph(new TranslationResult(TranslationManager.builder().build()));
		var ctx = new AnalysisContext(new File(markFilePaths.get(0)), graph);

		Map<String, Map<Integer, MarkIntermediateResult>> allResults = new TreeMap<>();
		for (MRule r : mark.getRules()) {
			MarkContextHolder markContextHolder = new MarkContextHolder();
			markContextHolder.getAllContexts().put(0, null); // add a dummy, so that we get exactly one result back for this context

			var ee = new ExpressionEvaluator(graph, mark, r, ctx, config, markContextHolder);

			Expression ensureExpr = r.getStatement().getEnsure().getExp();
			Map<Integer, MarkIntermediateResult> result = ee.evaluateExpression(ensureExpr);

			assertEquals(1, result.size());

			allResults.put(r.getName(), result);
		}

		allResults.forEach((key, value) -> {
			assertTrue(value.get(0) instanceof ConstantValue);

			MarkIntermediateResult inner = value.get(0);

			if (key.endsWith("true")) {
				assertEquals(true, ((ConstantValue) inner).getValue(), key);
			} else if (key.endsWith("false")) {
				assertEquals(false, ((ConstantValue) inner).getValue(), key);
			} else if (key.endsWith("fail")) {
				assertTrue(ConstantValue.isError(inner));
			} else {
				fail("Unexpected: Rule should have failed, but is " + inner + ": " + key);
			}
		});
	}

	@Test
	void testEquals() {
		test("equals.mark");
	}

	@Test
	void testLessThan() {
		test("lt.mark");
	}

	@Test
	void testGreaterThan() {
		test("gt.mark");
	}
}

File: src/test/java/de/fraunhofer/aisec/codyze/crymlin/RuleEnsureSemanticsTest.kt
package de.fraunhofer.aisec.codyze.crymlin

import de.fraunhofer.aisec.codyze.analysis.AnalysisContext
import de.fraunhofer.aisec.codyze.analysis.MarkContextHolder
import de.fraunhofer.aisec.codyze.analysis.MarkIntermediateResult
import de.fraunhofer.aisec.codyze.analysis.ServerConfiguration
import de.fraunhofer.aisec.codyze.analysis.TypestateMode
import de.fraunhofer.aisec.codyze.analysis.markevaluation.ExpressionEvaluator
import de.fraunhofer.aisec.codyze.analysis.resolution.ConstantValue
import de.fraunhofer.aisec.codyze.analysis.wpds.NFA
import de.fraunhofer.aisec.codyze.markmodel.MarkModelLoader
import de.fraunhofer.aisec.cpg.ExperimentalGraph
import de.fraunhofer.aisec.cpg.TranslationManager
import de.fraunhofer.aisec.cpg.TranslationResult
import de.fraunhofer.aisec.cpg.graph.graph
import de.fraunhofer.aisec.mark.XtextParser
import de.fraunhofer.aisec.mark.markDsl.MarkModel
import java.io.*
import java.util.*
import java.util.stream.Collectors
import org.junit.jupiter.api.*

internal class RuleEnsureSemanticsTest {
    @OptIn(ExperimentalGraph::class)
    private fun test(markFileEnding: String) {
        val markFilePaths =
            markModels
                .keys
                .stream()
                .filter { n: String -> n.endsWith(markFileEnding) }
                .collect(Collectors.toList())
        Assertions.assertEquals(1, markFilePaths.size)
        val mark = MarkModelLoader().load(markModels, markFilePaths[0])
        val config =
            ServerConfiguration.builder()
                .markFiles(markFilePaths[0])
                .typestateAnalysis(TypestateMode.NFA)
                .build()
        val graph = TranslationResult(TranslationManager.builder().build()).graph
        val ctx = AnalysisContext(File(markFilePaths[0]), graph)
        val allResults: MutableMap<String, Map<Int, MarkIntermediateResult>> = TreeMap()
        for (r in mark.rules) {
            val markContextHolder = MarkContextHolder()
            markContextHolder.allContexts[0] =
                null // add a dummy, so that we get exactly one result back for this context
            val ee = ExpressionEvaluator(graph, mark, r, ctx, config, markContextHolder)
            val ensureExpr = r.statement.ensure.exp
            val result = ee.evaluateExpression(ensureExpr)
            Assertions.assertEquals(1, result.size)
            allResults[r.name] = result
        }
        allResults.forEach { (key: String, value: Map<Int, MarkIntermediateResult>) ->
            Assertions.assertTrue(value[0] is ConstantValue)
            val inner = value[0]
            if (key.endsWith("true")) {
                Assertions.assertEquals(true, (inner as ConstantValue?)!!.value, key)
            } else if (key.endsWith("false")) {
                Assertions.assertEquals(false, (inner as ConstantValue?)!!.value, key)
            } else if (key.endsWith("fail")) {
                Assertions.assertTrue(ConstantValue.isError(inner))
            } else {
                Assertions.fail<Any>("Unexpected: Rule should have failed, but is $inner: $key")
            }
        }
    }

    @Test
    fun testEquals() {
        test("equals.mark")
    }

    @Test
    fun testLessThan() {
        test("lt.mark")
    }

    @Test
    fun testGreaterThan() {
        test("gt.mark")
    }

    companion object {
        private lateinit var markModels: HashMap<String, MarkModel>
        @BeforeAll
        @JvmStatic
        fun startup() {
            val resource =
                RuleEnsureSemanticsTest::class.java.classLoader.getResource(
                    "mark/rules/ensure/semantics/"
                )
            Assertions.assertNotNull(resource)
            val markFile = File(resource.file)
            Assertions.assertNotNull(markFile)
            var directoryContent =
                markFile.listFiles { current: File?, name: String -> name.endsWith(".mark") }
            if (directoryContent == null) {
                directoryContent = arrayOf(markFile)
            }
            Assertions.assertNotNull(directoryContent)
            Assertions.assertTrue(directoryContent.isNotEmpty())
            val parser = XtextParser()
            for (mf in directoryContent) {
                parser.addMarkFile(mf)
            }
            markModels = parser.parse()
            Assertions.assertFalse(markModels.isEmpty())
        }
    }
}
